File: functions/sequence_scheduler/template_engine.py
"""Simple template engine for email variable replacement."""

import copy
import os
import re
from typing import Any

# Base URL for unsubscribe links
API_BASE_URL = os.environ.get("API_BASE_URL", "https://api.envoy.app")
UNSUBSCRIBE_BASE_URL = f"{API_BASE_URL}/unsubscribe"


def replace_templates_in_text(text: str, replacements: dict[str, str]) -> str:
    """Replace {{variable}} placeholders in a string."""

    def replace_match(match: re.Match) -> str:
        var_name = match.group(1)
        return replacements.get(var_name, match.group(0))

    return re.sub(r"\{\{(\w+)\}\}", replace_match, text)


def replace_templates_in_blocks(
    builder_content: dict[str, dict[str, Any]],
    target_data: dict[str, Any],
    target_id: str,
) -> dict[str, dict[str, Any]]:
    """Replace {{variable}} placeholders in all block content.

    Supported variables:
    - {{first_name}}, {{last_name}}, {{company}}, {{title}}, {{email}}
    - {{unsubscribe_link}} - Full unsubscribe URL
    """
    replacements = {
        "first_name": target_data.get("first_name") or "",
        "last_name": target_data.get("last_name") or "",
        "company": target_data.get("company") or "",
        "title": target_data.get("title") or "",
        "email": target_data.get("email") or "",
        "unsubscribe_link": f"{UNSUBSCRIBE_BASE_URL}/{target_id}",
    }

    result = copy.deepcopy(builder_content)

    for block_id, block in result.items():
        block_type = block.get("type")
        props = block.get("data", {}).get("props", {})

        # Replace in text-based blocks
        if block_type in ("Text", "Heading", "Button") and "text" in props:
            props["text"] = replace_templates_in_text(props["text"], replacements)
        elif block_type == "Html" and "contents" in props:
            props["contents"] = replace_templates_in_text(props["contents"], replacements)

    return result

File: lib/template-engine.ts
/**
 * Simple template engine for email variable replacement.
 * Port of functions/sequence_scheduler/template_engine.py
 */
import { getEnv } from "@/lib/env";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type BlockMap = Record<string, Record<string, any>>;

function replaceTemplatesInText(
  text: string,
  replacements: Record<string, string>
): string {
  return text.replace(/\{\{(\w+)\}\}/g, (match, varName: string) => {
    return replacements[varName] ?? match;
  });
}

export function replaceTemplatesInBlocks(
  builderContent: BlockMap,
  targetData: Record<string, string | undefined | null>,
  targetId: string
): BlockMap {
  const baseUrl = getEnv().NEXT_PUBLIC_URL;

  const replacements: Record<string, string> = {
    first_name: targetData.first_name || "",
    last_name: targetData.last_name || "",
    company: targetData.company || "",
    title: targetData.title || "",
    email: targetData.email || "",
    unsubscribe_link: `${baseUrl}/unsubscribe/${targetId}`,
  };

  // Deep clone
  const result: BlockMap = JSON.parse(JSON.stringify(builderContent));

  for (const block of Object.values(result)) {
    const blockType = block.type;
    const props = block.data?.props;
    if (!props) continue;

    if (
      (blockType === "Text" || blockType === "Heading" || blockType === "Button") &&
      typeof props.text === "string"
    ) {
      props.text = replaceTemplatesInText(props.text, replacements);
    } else if (blockType === "Html" && typeof props.contents === "string") {
      props.contents = replaceTemplatesInText(props.contents, replacements);
    }
  }

  return result;
}
